ofnetSvcProxy.go

```diff
@@ -0,0 +1,662 @@
+        }
+
+	// Remove flows NAT'ed to this provider
+	for epIP, _ := range operEntry.provHdl[provIP].clientEPs {
+		for _, p := range operEntry.ports {
+			operEntry.delNATFlow(epIP, "Dst", &p)
+			operEntry.delNATFlow(epIP, "Src", &p)
+		}
+	}
+
+	// Remove provider from the loadbalancer pq
+	pqItem := operEntry.provHdl[provIP].pqHdl
+	operEntry.provPQ.RemoveItem(pqItem)
+	// remove the provider handle for this provider
+	delete(operEntry.provHdl, provIP)
+	log.Infof("Removed provider %s for serviceIP %s", provIP, svcIP)
+	return nil
+}
+
+// ProviderUpdate updates the list of providers of the service
+func (proxy *ServiceProxy)ProviderUpdate(svcName string, providers []string) {
+	log.Infof("ProviderUpdate %s %v", svcName, providers)
+	newProvs := make(map[string]bool)
+
+	for _, p := range providers {
+		newProvs[p] = true
+	}
+
+	pMap := Providers {
+			providers: newProvs,
+		}
+	// if we don't have the service spec yet, just save the provider
+	// map and return
+	sSpec, found := proxy.catalogue.svcMap[svcName]
+	if !found {
+		proxy.catalogue.provMap[svcName] = pMap
+		log.Debugf("Service %s -- no spec yet", svcName)
+		return
+	}
+
+	// if the service is not created, just use the new map and 
+	// add the service
+	svcIP := sSpec.IpAddress
+	_, found = proxy.operState[svcIP]
+	if !found && len(providers) == 0 {
+		log.Debugf("Service %s -- no providers", svcName)
+		return
+	}
+
+	if !found {
+		proxy.catalogue.provMap[svcName] = pMap
+		proxy.addService(svcName)
+		return
+	}
+
+	currProvs := proxy.catalogue.provMap[svcName]
+	proxy.catalogue.provMap[svcName] = pMap
+
+	// if the new provider list is empty, delete the service
+	if len(providers) == 0 {
+		log.Infof("No providers for service %s, deleting", svcName)
+		proxy.delService(svcName)
+		return
+	}
+
+	// Add any new providers first
+	for _, p := range providers {
+		_, found = currProvs.providers[p]
+		if !found {
+			proxy.addProvider(sSpec.IpAddress, p)
+		}
+	}
+
+	// Delete any providers that disappeared
+	for p, _ := range currProvs.providers {
+		_, found = newProvs[p]
+		if !found {
+			proxy.delProvider(sSpec.IpAddress, p)
+		}
+	}
+}
+
+// NewServiceProxy Creates a new service proxy
+func NewServiceProxy() *ServiceProxy {
+	svcProxy := new(ServiceProxy)
+
+	// initialize
+	svcProxy.catalogue.svcMap = make(map[string]ServiceSpec)
+	svcProxy.catalogue.provMap = make(map[string]Providers)
+	svcProxy.operState = make(map[string]*proxyOper)
+
+	return svcProxy
+}
+
+// Handle switch connected notification
+func (proxy *ServiceProxy) SwitchConnected(sw *ofctrl.OFSwitch) {
+	// Keep a reference to the switch
+	proxy.ofSwitch = sw
+
+	log.Infof("Switch connected(svcProxy).")
+}
+
+// Handle switch disconnected notification
+func (proxy *ServiceProxy) SwitchDisconnected(sw *ofctrl.OFSwitch) {
+	// FIXME: ??
+}
+
+// DelEndpoint handles an endpoint delete
+func (proxy *ServiceProxy) DelEndpoint(endpoint *OfnetEndpoint) {
+	epIP := endpoint.IpAddr.String()
```

Comment:
Yes, we will get the provider update for that. We are likely to get that before the endpoint is deleted.